pkg/providers/vsphere/validator.go

```diff
@@ -385,3 +419,234 @@ func (v *Validator) validateControlPlaneIpUniqueness(spec *Spec) error {
+		ctx,
+		host,
+		vuc.EksaVsphereCPUsername,
+		vuc.EksaVsphereCPPassword,
+		spec.datacenterConfig.Spec.Insecure,
+		datacenter,
+	)
+	if err != nil {
+		return err
+	}
+
+	return v.validatePrivs(ctx, ras, vsc)
+}
+
+func (v *Validator) validateCSIUserPrivs(ctx context.Context, spec *Spec, vuc *config.VSphereUserConfig) error {
+	ras := []RequiredAccess{
+		{ // CNS-SEARCH-AND-SPBM role
+			objectType:   vsphereTypeFolder,
+			privsContent: config.VSphereCnsSearchAndSpbmPrivsFile,
+			path:         vsphereRootPath,
+		},
+	}
+
+	machineConfigs, err := v.collectSpecMachineConfigs(ctx, spec)
+	if err != nil {
+		return err
+	}
+
+	var newRas []RequiredAccess
+	for _, mc := range machineConfigs {
+		newRas = []RequiredAccess{
+			{ // CNS-Datastore role
+				objectType:   vsphereTypeDatastore,
+				privsContent: config.VSphereCnsDatastorePrivsFile,
+				path:         mc.Spec.Datastore,
+			},
+
+			{ // CNS-VM role
+				objectType:   vsphereTypeFolder,
+				privsContent: config.VSphereCnsVmPrivsFile,
+				path:         mc.Spec.Folder,
+			},
+			// CNS-HOST-CONFIG-STORAGE role
+			{
+				objectType:   vsphereTypeDatastore,
+				privsContent: config.VSphereCnsHostConfigStorageFile,
+				path:         mc.Spec.Datastore,
+			},
+		}
+		ras = append(ras, newRas...)
+	}
+
+	host := spec.datacenterConfig.Spec.Server
+	datacenter := spec.datacenterConfig.Spec.Datacenter
+
+	vsc, err := v.vSphereClientBuilder.Build(
+		ctx,
+		host,
+		vuc.EksaVsphereCSIUsername,
+		vuc.EksaVsphereCSIPassword,
+		spec.datacenterConfig.Spec.Insecure,
+		datacenter,
+	)
+	if err != nil {
+		return err
+	}
+
+	return v.validatePrivs(ctx, ras, vsc)
+}
+
+func (v *Validator) validateCPUserPrivs(ctx context.Context, spec *Spec, vuc *config.VSphereUserConfig) error {
+	// CP role just needs read only
+	privObjs := []RequiredAccess{
+		{
+			objectType:   vsphereTypeFolder,
+			privsContent: config.VSphereReadOnlyPrivs,
+			path:         vsphereRootPath,
+		},
+	}
+
+	host := spec.datacenterConfig.Spec.Server
+	datacenter := spec.datacenterConfig.Spec.Datacenter
+
+	vsc, err := v.vSphereClientBuilder.Build(
+		ctx,
+		host,
+		vuc.EksaVsphereCPUsername,
+		vuc.EksaVsphereCPPassword,
+		spec.datacenterConfig.Spec.Insecure,
+		datacenter,
+	)
+	if err != nil {
+		return err
+	}
+
+	return v.validatePrivs(ctx, privObjs, vsc)
+}
+
+func (v *Validator) validatePrivs(ctx context.Context, privObjs []RequiredAccess, vsc VSphereClient) error {
+	var missingPrivs []string
+	var err error
+
+	for _, obj := range privObjs {
+		path := obj.path
+		privsContent := obj.privsContent
+		t := obj.objectType
+		missingPrivs, err = v.getMissingPrivs(ctx, vsc, path, t, privsContent, vsc.Username())
+		if err != nil {
+			return err
+		} else if len(missingPrivs) > 0 {
+			return fmt.Errorf("User %s missing privileges on %s: %v", vsc.Username(), path, missingPrivs)
+		}
+	}
+
+	return nil
+}
+
+func diffPrivs(requiredPrivs []string, hasPrivs []string) []string {
```

Comment:
Diff makes it sound like we are running a diff, but this is almost a subset of a diff imo
```suggestion
func checkRequiredPrivs(requiredPrivs []string, hasPrivs []string) []string {
```

Reply:
committed locally 👍